server/service/software_title_icons.go

```diff
@@ -0,0 +1,276 @@
+}
+
+func (r getSoftwareTitleIconsResponse) HijackRender(ctx context.Context, w http.ResponseWriter) {
+	if r.Err != nil {
+		return
+	}
+
+	w.Header().Set("Content-Type", r.ContentType)
+	w.Header().Set("Content-Disposition", fmt.Sprintf(`inline; filename="%s"`, r.Filename))
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", r.Size))
+
+	_, _ = w.Write(r.ImageData)
+}
+
+func getSoftwareTitleIconsEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (fleet.Errorer, error) {
+	req := request.(*getSoftwareTitleIconsRequest)
+
+	if req.TeamID == nil {
+		return getSoftwareTitleIconsResponse{Err: &fleet.BadRequestError{Message: "team_id is required"}}, nil
+	}
+	if req.TitleID == 0 {
+		return getSoftwareTitleIconsResponse{Err: &fleet.BadRequestError{Message: "invalid title_id"}}, nil
+	}
+
+	iconData, size, filename, err := svc.GetSoftwareTitleIcon(ctx, *req.TeamID, req.TitleID)
+	if err != nil {
+		var vppErr *fleet.VPPIconAvailable
+		if errors.As(err, &vppErr) {
+			// 302 redirect to vpp app IconURL
+			return getSoftwareTitleIconsRedirectResponse{RedirectURL: vppErr.IconURL}, nil
+		}
+		return getSoftwareTitleIconsResponse{Err: err}, nil
+	}
+
+	return getSoftwareTitleIconsResponse{
+		ImageData:   iconData,
+		ContentType: "image/png", // only type of icon we currently allow
+		Filename:    filename,
+		Size:        size,
+	}, nil
+}
+
+func (svc *Service) GetSoftwareTitleIcon(ctx context.Context, teamID uint, titleID uint) ([]byte, int64, string, error) {
+	// skipauth: No authorization check needed due to implementation returning
+	// only license error.
+	svc.authz.SkipAuthorization(ctx)
+
+	return nil, 0, "", fleet.ErrMissingLicense
+}
+
+type putSoftwareTitleIconRequest struct {
+	TitleID  uint  `url:"title_id"`
+	TeamID   *uint `query:"team_id"`
+	File     *multipart.FileHeader
+	Hash     *string
+	Filename *string
+}
+
+type putSoftwareTitleIconResponse struct {
+	Err     error  `json:"error,omitempty"`
+	IconUrl string `json:"icon_url,omitempty"`
+}
+
+func (r putSoftwareTitleIconResponse) Error() error {
+	return r.Err
+}
+
+func (putSoftwareTitleIconRequest) DecodeRequest(ctx context.Context, r *http.Request) (interface{}, error) {
+	urlVars := mux.Vars(r)
+	titleID, ok := urlVars["title_id"]
+	if !ok {
+		return nil, &fleet.BadRequestError{Message: "title_id is required"}
+	}
+	titleIDUint64, err := strconv.ParseUint(titleID, 10, 64)
+	if err != nil {
+		return nil, &fleet.BadRequestError{Message: "invalid title_id"}
+	}
+	teamID := r.URL.Query().Get("team_id")
+	if teamID == "" {
+		return nil, &fleet.BadRequestError{Message: "team_id is required"}
+	}
+	teamIDUint64, err := strconv.ParseUint(teamID, 10, 64)
+	if err != nil {
+		return nil, &fleet.BadRequestError{Message: "invalid team_id"}
+	}
+	teamIDUint := uint(teamIDUint64)
+
+	decoded := putSoftwareTitleIconRequest{
+		TitleID: uint(titleIDUint64),
+		TeamID:  &teamIDUint,
+	}
+
+	err = r.ParseMultipartForm(512 * units.MiB)
```

Comment:
```suggestion
	err = r.ParseMultipartForm(6 * units.MiB)
```

Our max size for PNGs is 1024x1024px, and an rgba bitmap at that resolution would be 4 MiB. So we can be a lot tighter on what we parse here.